# **Intro to SiS1 Labs**

The goal of the labs of Signal and Systems 1 is to gain knowledge on the course’s topics by solving a specific signal processing problem. By the end of the course, you should be able to imitate, by synthesis, the sound produced by a plucked string instrument using the knowledge acquired in the theory and seminar classes. This will be a cumulative process, in each lab new techniques will be applied in order to improve the synthesis. First let's show what we will achieve by the end of the course. For each lab we need to import some auxiliary functions needed to load the audio files and listen to them:


In [1]:
from util import load_audio, save_audio, plot_signals
from IPython.display import Audio

**Reference signal**

Now let's load and listen to a reference signal like the one you will use:

In [2]:
reference, fs = load_audio('audio/reference.wav')
Audio(reference, rate=fs)

**Synthesis**

Now we can load and listen to a synthesised signal like the one you will obtain and that tries to imitate the reference signal. This is an audio signal created from scratch only using the knowledge gained during the course.

In [3]:
synthesis, fs = load_audio('audio/synthesis.wav')
Audio(synthesis, rate=fs)

---

## **Lab 1**
In this lab you will choose your reference sound to be used for all the labs and start imitating it with a single sinusoid.

## **Exercises**

**1. Choose a reference sound**

1.1. Find a recording of a single plucked string sound. Be creative, do not choose “guitar” by default, nor copy a sound from another group. You should use [Freesound](https://freesound.org/) to find the recordings (Freesound is a huge collaborative database of audio snippets released under Creative Commons licenses created and maintained by researchers of the Music Technology Group of the UPF). Try to find a sound with the best possible quality. The ideal sound format should be: uncompress (such as.wav, .aiff, but no .mp3), mono (one channel) and with a sampling rate of  44100Hz. The sound file should also contain a single note. You can start by downloading a sound from freesound and modify it with Audacity to the right format and content.

[Put here the freesound link to the selected sound recording]

https://freesound.org/s/392396/

1.2. Create a github repository and upload your edited audio file there:

1) Sing up to github using your UPF gmail account: https://github.com/

2) Create a repository called **sis1_groupX** where X is your group number: https://docs.github.com/en/get-started/quickstart/create-a-repo

3) Upload your edited audio file to the repository


1.3. Now you can clone your repository and access the audio file locally. Change the following code (user name and file name) to clone the repository and load the audio file.

In [4]:
#!git clone https://github.com/guillemmarti06/sis1_group103.git
filepath = "./sis1_group103/audio_la.wav"
ref, fs = load_audio(filepath)

Note that `load_audio` function returns two variables: the audio signal (as a one dimensional array of floating point numbers), and the sampling rate (as a integer number). Now we can plot the audio signal:

In [5]:
plot_signals(ref, fs)

You can use the zoom in tool to see more details.

You can also use Audio widget to listen the audio signal:

In [6]:
Audio(ref, rate=fs)

**2. Measure the signal period and fundamental frequency**

2.1. Measure the period length, in seconds, and the fundamental frequency, in Hz, of the sound you choose. You can use Audacity to measure the period by zooming into a stable portion of the sound. Then compute its inverse to find the fundamental frequency. Find the closest note (note name) of the measured frequency (use google).

Zooming in with audacity and finding two consecutive points, we can see how the period is approximately T = 0,00227 seconds. 
Then, computing the inverse, we find that the fundamental frequency is f = 1 / T = 1 / 0,00227 = 440,5 Hz.
By using google, I found that the closest note is A4.

2.2. Calculate the period from the frequency value using Python:


In [7]:
# paste the values 
f = 440.5  
T = 1 / f

print(f"The period is {T:.6f} seconds")

The period is 0.002270 seconds


2.3. Plot three periods of the signal using the `plot_signals` function and selecting `t_start` and `t_end` arguments accordingly. Try to get the first periods of the audio signal but avoiding the attack section.

In [8]:
# define the period 
T = 1 / 440.5  

# choose a constant region
t_start = 0.2                 
t_end = t_start + 3 * T       

# Plot the signal
plot_signals(ref, fs, t_start=t_start, t_end=t_end)

2.4. Measure the period of the signal manually by identifying the begining and ending of a periodic cycle. Note that when you hover the mouse over one of the points in the plot, you can see the time and amplitude values. For instance, you can check the cross by zero.

The Period measured manually is T = 0,2033807 - 0,2011358 = 0,0022449 seconds

2.5. Do the measured (2.4) and theoretical (2.2) periods coincide? If not, explain why.

They are quite close to each other. In Audacity, I tried to be as precise as possible by zooming in a lot. The values are not exactly the same; in section 2.2 we found 0.00227 s, while here (more precisely) we obtained 0.0022449 s, which is slightly smaller. However, in general terms, both results are very similar, especially considering how small these values are.

**3. Generating a sinusoid**

3.1. Create a sinusoid of the same duration than the reference signal and same frequency than the fundamental frequency. Plot it along with the audio signal of the reference recording. Plot the same time segment that 2.3. Note that the sampling rate should be the same for both signals. Try to find the values of Amplitude and initial Phase that make the two signal segments plotted to match as close as possible (do not attempt to imitate the whole signal). Matching the phase is a bit tricky.



In [ ]:
import numpy as np

# parameters
fs = fs              # same sampling rate as the reference (requested)
f0 = 440.5           # frequency from 2.1
A = 0.5              # adjust amplitude to match 
phi = np.pi/4             
t = np.arange(0, len(ref)) / fs   # same duration as reference

# create sinusoid
x = A * np.cos(2 * np.pi * f0 * t + phi) # 2pif = w

# plot both signals over the same 3-period as in 2.3
t_start = 0.2
t_end = t_start + 3 * (1 / f0)

plot_signals([ref, x], fs, t_start=t_start, t_end=t_end, name=['reference signal', 'generated sinusoid'])


3.2. What are the main differences between the two signals.

We can see that the reference signal (blue) is more complex, while the generated sinusoid (orange) is smooth and perfectly periodic. Moreover, the reference has small variations and additional oscillations because it contains many harmonics, and obviously a decaying amplitude (because it’s a real sound produced by an instrument).
The sinusoid only represents the fundamental frequency, with constant amplitude and no harmonics or decay.
Therefore, the sinusoid only approximates one simple component of the real sound.

3.3. Listen to the synthesized signal usign the Audio widget. Sounds natural?

The synthesized signal roughly imitates the “La” sound, but it maintains the same intensity throughout. What I mean with that; the original sound starts strong and gradually decreases in amplitude over time, producing a natural decay. Because of this, the pure sinusoid sounds artificial, it lacks the variations in amplitude and the harmonics that make real instrument sounds richer and more natural. So, in general, it kind of imitates the sound, but it's way far from sounding as natural as a real instrument.

In [14]:
Audio(x, rate=fs)

3.4. Save the synthesized signal using the 'save_audio` function:

In [16]:
#we save it as synthesized_la in order to distinguish from the ref one
save_audio('sis1_group103/synthesized_la.wav', x, fs)